File: kotlin/sdk/src/commonMain/kotlin/com/basecamp/fizzy/Pagination.kt
package com.basecamp.fizzy

/**
 * Metadata about a paginated list response.
 *
 * Fizzy does not return X-Total-Count headers, so [truncated] indicates
 * whether more results exist beyond what was returned.
 */
data class ListMeta(
    /** True when results were truncated (by maxItems or page safety cap). */
    val truncated: Boolean,
)

/**
 * Options for controlling pagination behavior.
 */
data class PaginationOptions(
    /**
     * Maximum number of items to return across all pages.
     * When null or 0, all pages are fetched.
     */
    val maxItems: Int? = null,
)

/**
 * A list of results with pagination metadata.
 *
 * Delegates to `List<T>` so it's fully compatible with all collection operations
 * (`.forEach()`, `.map()`, `.size`, indexing, etc.). Additional metadata is
 * accessible via the [meta] property.
 *
 * ```kotlin
 * val cards = account.cards.list(boardId)
 * println("Got ${cards.size} cards (truncated: ${cards.meta.truncated})")
 * cards.forEach { println(it.title) }
 * ```
 */
class ListResult<T>(
    private val items: List<T>,
    /** Pagination metadata (truncation status). */
    val meta: ListMeta,
) : List<T> by items {

    override fun toString(): String = "ListResult(size=$size, meta=$meta)"

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ListResult<*>) return false
        return items == other.items && meta == other.meta
    }

    override fun hashCode(): Int = 31 * items.hashCode() + meta.hashCode()
}

/**
 * Extracts the `rel="next"` URL from a Link header.
 * Returns null if no next link exists.
 *
 * Example: `<https://fizzy.do/api/cards?page=2>; rel="next"` -> the URL
 */
internal fun parseNextLink(linkHeader: String?): String? {
    if (linkHeader.isNullOrBlank()) return null
    for (part in linkHeader.split(",")) {
        val trimmed = part.trim()
        if (trimmed.contains("""rel="next"""")) {
            val start = trimmed.indexOf('<')
            val end = trimmed.indexOf('>')
            if (start >= 0 && end > start) {
                return trimmed.substring(start + 1, end)
            }
        }
    }
    return null
}

/**
 * Validates that two URLs share the same origin (scheme + host + port).
 * Used to prevent SSRF via poisoned Link headers.
 */
internal fun isSameOrigin(url1: String, url2: String): Boolean {
    val origin1 = extractOrigin(url1) ?: return false
    val origin2 = extractOrigin(url2) ?: return false
    return origin1 == origin2
}

/** Extracts scheme://host:port from a URL string. */
private fun extractOrigin(url: String): String? {
    val schemeEnd = url.indexOf("://")
    if (schemeEnd < 0) return null
    val afterScheme = schemeEnd + 3
    // Find end of authority (host:port) -- next / or end of string
    val pathStart = url.indexOf('/', afterScheme)
    val authority = if (pathStart < 0) url.substring(afterScheme) else url.substring(afterScheme, pathStart)
    return url.substring(0, schemeEnd) + "://" + authority
}

/**
 * Parses the Retry-After header value.
 * Supports integer seconds format.
 * Returns null if the header is missing or cannot be parsed.
 */
internal fun parseRetryAfter(value: String?): Int? {
    if (value.isNullOrBlank()) return null
    val seconds = value.trim().toIntOrNull()
    return if (seconds != null && seconds > 0) seconds else null
}

File: swift/Sources/Fizzy/Pagination.swift
import Foundation

/// Metadata about a paginated list response.
public struct ListMeta: Sendable, Equatable {
    /// True when results were truncated (by `maxItems` or page safety cap).
    public let truncated: Bool

    public init(truncated: Bool = false) {
        self.truncated = truncated
    }
}

/// Options for controlling pagination behavior.
public struct PaginationOptions: Sendable {
    /// Maximum number of items to return across all pages.
    /// When nil or 0, all pages are fetched.
    public let maxItems: Int?

    public init(maxItems: Int? = nil) {
        self.maxItems = maxItems
    }
}

/// A paginated list result that conforms to `RandomAccessCollection`.
///
/// Acts like a Swift `Array` -- supports `for-in`, `.count`, subscripting,
/// `.map()`, `.filter()`, and all other collection operations. The `.meta`
/// property provides pagination metadata.
///
/// ```swift
/// let cards = try await client.cards.list(boardId: 1)
/// print("Showing \(cards.count) cards")
/// for card in cards { print(card.title) }
/// let titles = cards.map(\.title)
/// ```
public struct ListResult<Element: Sendable>: Sendable {
    /// The underlying items.
    public let items: [Element]
    /// Pagination metadata.
    public let meta: ListMeta

    /// Creates a new list result.
    public init(_ items: [Element], meta: ListMeta) {
        self.items = items
        self.meta = meta
    }

    /// Creates an empty list result.
    public init() {
        self.items = []
        self.meta = ListMeta()
    }
}

// MARK: - RandomAccessCollection

extension ListResult: RandomAccessCollection {
    public typealias Index = Int

    public var startIndex: Int { items.startIndex }
    public var endIndex: Int { items.endIndex }

    public subscript(position: Int) -> Element {
        items[position]
    }
}

// MARK: - Pagination Utilities

/// Parses the next URL from a Link header.
///
/// Looks for `rel="next"` in the header value.
///
/// - Parameter linkHeader: The Link header value.
/// - Returns: The URL for the next page, or nil if not found.
func parseNextLink(_ linkHeader: String?) -> String? {
    guard let linkHeader, !linkHeader.isEmpty else { return nil }

    for part in linkHeader.split(separator: ",") {
        let trimmed = part.trimmingCharacters(in: .whitespaces)
        if trimmed.contains("rel=\"next\"") {
            guard let start = trimmed.firstIndex(of: "<"),
                  let end = trimmed.firstIndex(of: ">"),
                  start < end
            else { continue }
            return String(trimmed[trimmed.index(after: start)..<end])
        }
    }
    return nil
}

/// Resolves a possibly-relative URL against a base URL.
///
/// If target is already absolute, it is returned unchanged.
func resolveURL(base: String, target: String) -> String {
    guard let baseURL = URL(string: base) else { return target }
    guard let resolved = URL(string: target, relativeTo: baseURL) else { return target }
    return resolved.absoluteString
}

/// Checks whether two absolute URLs share the same origin (scheme + host + port).
func isSameOrigin(_ a: String, _ b: String) -> Bool {
    guard let urlA = URLComponents(string: a),
          let urlB = URLComponents(string: b)
    else { return false }

    return urlA.scheme == urlB.scheme
        && urlA.host == urlB.host
        && (urlA.port ?? defaultPort(for: urlA.scheme)) == (urlB.port ?? defaultPort(for: urlB.scheme))
}

private func defaultPort(for scheme: String?) -> Int? {
    switch scheme {
    case "https": 443
    case "http": 80
    default: nil
    }
}
